src/fhirtypes/common.ts

```diff
@@ -1601,3 +1601,40 @@ export function findImposeProfiles(sd: any): string[] | undefined {
     return imposeProfiles;
   }
 }
+
+// Used to check if the entity used in Canonical() references a contained resource.
+// Checks a list of validated rules at any contained path for a matching value from the
+// Canonical() keyword and returns the matching resource's id.
+export function getMatchingContainedReferenceId(
+  value: string,
+  containedResources: { pathParts: PathPart[]; assignedValue: any }[]
+) {
+  const matchingContainedResource = containedResources.find(
+    r =>
+      r.assignedValue?.url === value ||
+      r.assignedValue?.name === value ||
+      r.assignedValue?.id === value
+  );
+  const matchingContainedResourceId = containedResources.find(
+    r => r.pathParts.slice(-1)[0].base === 'id'
+  );
+  const hasMatchingContainedResourceId = containedResources.some(
+    r => r.pathParts.slice(-1)[0].base === 'id' && r.assignedValue === value
+  );
+  const hasMatchingContainedResourceName = containedResources.some(
+    r => r.pathParts.slice(-1)[0].base === 'name' && r.assignedValue === value
+  );
+  const hasMatchingContainedResourceUrl = containedResources.some(
+    r => r.pathParts.slice(-1)[0].base === 'url' && r.assignedValue === value
+  );
```

Comment:
This implementation can get the incorrect result if you have more than one contained resource that has its elements assigned directly in the containing resource. Here's an example:
```
Instance: MyObservation
InstanceOf: Observation
Usage: #example
* status = #draft
* code = #123
* contained[0]
  * resourceType = "Observation"
  * id = "obs-one"
  * status = #draft
  * code = #111
* contained[1]
  * resourceType = "Observation"
  * id = "obs-two"
  * status = #draft
  * code = #222
* valueString = Canonical(obs-two)
```
The value of `valueString` will be `#obs-one`, since the rule on `contained[0].id` is the one that is found for `matchingContainedResourceId`.

Reply:
Good catch! I'll fix that, thanks!

Reply:
Fixed in [e4837bd](https://github.com/FHIR/sushi/pull/1405/commits/e4837bd810dd83692f6be497b890f27510ceb6e6).